### ISSUE TITLE
[BUG] Issue with Notepad++ Not Supporting Long Paths

### ISSUE THREAD

**Issue by softmgr** [notepad-plus-plus/notepad-plus-plus#15828]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description of the Issue

In older versions of Windows system APIs, file paths are limited to a maximum length of 260 characters:
```
#define MAX_PATH          260
```

When cloning certain Git repositories from Linux that contain files with long path names, errors may occur when Windows tries to access these paths using standard APIs. This introduces the need for Unicode versions of the Windows API.

For scenarios requiring long path support, you can prefix the local Windows path with `\\?\` or `\\?\UNC\`. For example:
```
\\?\D:\LongPath\LongPath\LongPath\...
```

The Unicode version of the path allows a maximum total path length of up to 32,767 characters:
```
#define MAX_UNICODE_PATH          32767
```

Enabling Long Paths in Windows 10 Version 1607 and Later

To enable long paths in Windows 10 version 1607 or higher, two conditions must be met:

	1.	A registry entry must exist at:
Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem\LongPathsEnabled
(Type: REG_DWORD) and must be set to 1:
```
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem]
"LongPathsEnabled"=dword:00000001
```

	2.	The application manifest must include the longPathAware element:
```
<application xmlns="urn:schemas-microsoft-com:asm.v3">
    <windowsSettings xmlns:ws2="http://schemas.microsoft.com/SMI/2016/WindowsSettings">
        <ws2:longPathAware>true</ws2:longPathAware>
    </windowsSettings>
</application>
```

### Steps To Reproduce

1.  Open a file with long path.

### Current Behavior

Could not open file.

### Expected Behavior

Can open these files.

### Debug Information

```shell
Notepad++ v8.7.1   (64-bit)
Build time : Oct 31 2024 - 00:48:56
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : 
Admin mode : ON
Local Conf mode : OFF
Cloud Config : OFF
Periodic Backup : ON
OS Name : Windows 7 Ultimate (64-bit)
OS Version : 23H2
OS Build : 7600.4460
Current ANSI codepage : 1252
Plugins : 
    mimeTools (3.1)
    NppConverter (4.6)
    NppExport (0.4)
```

### Anything else?

Directory-related APIs
```
	CreateDirectoryW
	CreateDirectoryExW
	GetCurrentDirectoryW
	RemoveDirectoryW
	SetCurrentDirectoryW
```

File-related APIs
```
	CopyFileW
	CopyFile2
	CopyFileExW
	CreateFileW
	CreateFile2
	CreateHardLinkW
	CreateSymbolicLinkW
	DeleteFileW
	FindFirstFileW
	FindFirstFileExW
	FindNextFileW
	GetFileAttributesW
	GetFileAttributesExW
	SetFileAttributesW
	GetFullPathNameW
	GetLongPathNameW
	MoveFileW
	MoveFileExW
	MoveFileWithProgressW
	ReplaceFileW
	SearchPathW
	FindFirstFileNameW
	FindNextFileNameW
	FindFirstStreamW
	FindNextStreamW
	GetCompressedFileSizeW
	GetFinalPathNameByHandleW
```

**Comment by alankilborn** [comment]
> I have searched the existing issues

I **_seriously_** don't think you did.  If you did, you would have seen at least one of these:

https://github.com/notepad-plus-plus/notepad-plus-plus/issues/4078
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/4483
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/5061
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8501
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9181
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9946
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/10493   <-- why is this 404?
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11754
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/12777
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/14015
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/15649
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13953
https://community.notepad-plus-plus.org/topic/20827/can-t-open-files-with-103-char-filenames-259-char-paths
https://community.notepad-plus-plus.org/topic/18372/long-paths-support

**Comment by softmgr** [comment]
@donho 
I have modified some code and tested it successfully.
However, after Notepad++ became compatible with long paths, some strange issues started to appear.
These include crashes caused by third-party plugins that do not support long paths, as well as problems arising from Windows’ incomplete support for long paths.
To achieve perfect compatibility with long paths, there is still a lot of work to be done.

![snap-1](https://github.com/user-attachments/assets/8644f8e1-3110-4960-9359-fd266a2de20c)

**Comment by xomx** [comment]
@donho 
> @xomx
> IIRC, you have talked about breaking path length limit in the past.
> Do you think the solution is so simple?

Of course not.

https://github.com/notepad-plus-plus/notepad-plus-plus/issues/15649#issuecomment-2371846621

I'd like to see the long filenames support in N++ too but one 1st has to:
- review & rewrite all the N++ code where MAX_PATH static-size buffers are used and replace them e.g. by the std::wstring unlimited buffers (where are the static-size buffers really needed because of the WINAPI requirement, there always create a temp buffer of the correct size for a string in use)
- create new N++ plugin API (MAX_PATH limit free) but leave the old one in N++ for compatibility with the old plugins

**Comment by iscshuai** [comment]
This is due to the maximum path length limitation of Windows. In the Windows API, the maximum length of a path is MAX_PATH, defined as 260 characters. Once this limit is exceeded, Notepad++ cannot open the files in the path. 
One of the solutions can be referred to:
https://learn.microsoft.com/zh-cn/windows/win32/fileio/maximum-file-path-limitation?tabs=powershell

**Comment by softmgr** [comment]
> This is due to the maximum path length limitation of Windows. In the Windows API, the maximum length of a path is MAX_PATH, defined as 260 characters. Once this limit is exceeded, Notepad++ cannot open the files in the path. 
> One of the solutions can be referred to:
> https://learn.microsoft.com/zh-cn/windows/win32/fileio/maximum-file-path-limitation?tabs=powershell


@iscshuai I have tested this on the latest versions of Windows 10 (22H2) and Windows 11 (24H2), and even when `LongPathsEnabled` is enabled in the registry, the system's File Explorer still cannot handle long paths correctly. This is a long-standing bug in File Explorer that has persisted since the Windows 7 era and remains unfixed.

Specifically, the issue is as follows:
When navigating into a long path folder in File Explorer and then double-clicking a TXT file, an error message appears stating that the file cannot be opened. The same error occurs when using the right-click "Open" context menu option. However, if you use the "Open with" dialog and choose any application that supports long paths, such as Notepad, the file can be opened normally.

Additionally, I have tested that the same double-click or right-click "Open" action on a TXT file works correctly in the Total Commander file manager, which does not have this bug.

Currently, Notepad++ does not support long paths, even if `LongPathsEnabled` is enabled in the registry. I have modified Notepad++ myself to basically fully support long paths. This requires defining an additional constant:

```
#define MAX_UNICODE_PATH 32767
```

And implementing two long path conversion functions like the following:
```
bool addLongPathPrefix(std::wstring& path)
{
	if (path.size() > MAX_PATH) {
		const std::wstring longPathPrefix = L"\\\\?\\";

		if (path.compare(0, longPathPrefix.size(), longPathPrefix) != 0) {
			path = longPathPrefix + path;
			return true;
		}
	}
	return false;
}

bool addLongPathPrefix(wchar_t* path, const size_t bufferSize/* = MAX_UNICODE_PATH*/)
{
	const size_t pathLength = (path != nullptr) ? wcslen(path) : 0;
	if (pathLength > MAX_PATH) {
		const wchar_t* longPathPrefix = L"\\\\?\\";
		const size_t prefixLength = wcslen(longPathPrefix);

		if (wcsncmp(path, longPathPrefix, prefixLength) == 0) {
			return false;
		}

		if (pathLength + prefixLength + 1 > bufferSize) {
			return false;
		}

		for (size_t i = pathLength; i != static_cast<size_t>(-1); --i) {
			path[i + prefixLength] = path[i];
		}

		wcsncpy(path, longPathPrefix, prefixLength);
		return true;
	}
	return false;
}
```

**Comment by mgroeber9110** [comment]
Would it be possible to first add limited support for long paths using the Windows `GetShortPathName` API function (see [here](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getshortpathnamew))?

It appears that `GetShortPathName` can handle long pathnames as input, as long they are prefixed with the `\\?\` string.

The idea would be: If NPP is launched with a long path name on the command line, first attempt to translate this path using `GetShortPathName` into a path using only 8+3 elements. If this path fits the 259 character limit, use it for all further processing, otherwise fail with an error.

When trying this briefly with a path translated by hand (using "echo %~s1" in a batch file), I found that NPP will translate this back to the long path, so there may be additionally changes needed.

However, this may also offer a partial way of maintaining backwards compatibility to old plugins: those using the non-long version of the API would still be able to access files whose shortened path fits within the limit.

---

Of course, this still restricts use cases where files in deep directory structures or with very long names are accessed in other way, e.g. by navigating through the Open dialog, but it would unblock the most common (?) case of double-clicking of a file inside a deep directory. For example, this is the most common use-case for me when trying to view/edit files in WinSCP that live in deeply nested file system. 

---

Addendum: to experiment with the behavior of NPP with 8+3 pathnames, I have written a simple cmd file `nppwrap.cmd` with this line:

    start "" "\Program Files\Notepad++\notepad++.exe" %~s1

This can be associated with file types that are commonly found in deep file trees (in my case, .error and .log), so that double-clicking on the file will open NPP with the shortened name. I have found that, if the original path was actually short enough, NPP will expand the name back to its full form, while it remains shortened for longer paths (perhaps as part of file name canonicalization by Windows) - which is almost as good as it can get without fully supporting long names.

[Edit 18-Nov-2025: fixed start command so Notepad++ is directly launched rather than with a file association.]

**Comment by BrBill** [comment]
> [@iscshuai](https://github.com/iscshuai) I have tested this on the latest versions of Windows 10 (22H2) and Windows 11 (24H2), and even when `LongPathsEnabled` is enabled in the registry, the system's File Explorer still cannot handle long paths correctly. This is a long-standing bug in File Explorer that has persisted since the Windows 7 era and remains unfixed.
> 
> Specifically, the issue is as follows: When navigating into a long path folder in File Explorer and then double-clicking a TXT file, an error message appears stating that the file cannot be opened. The same error occurs when using the right-click "Open" context menu option. However, if you use the "Open with" dialog and choose any application that supports long paths, such as Notepad, the file can be opened normally.

In latest Win11 yesterday, I tried double-clicking on very long paths in Explorer, and no error is generated. One of the following things happens I double-click on a file in this situation:
- N++ launches but doesn't open the file
- Nothing happens

I haven't been able to determine a reliable set of circumstances that will make one or the other thing happen yet.

**Comment by chendrp** [comment]
@softmgr  We've encountered this issue as well. Could you advise on how to modify the source code to add these two functions, including the specific file directory paths? Alternatively, would it be possible for you to submit a pull request to fix this issue? Thank you.😁

**Comment by softmgr** [comment]
> [@softmgr](https://github.com/softmgr) We've encountered this issue as well. Could you advise on how to modify the source code to add these two functions, including the specific file directory paths? Alternatively, would it be possible for you to submit a pull request to fix this issue? Thank you.😁


I solved this issue a long time ago for my own private enhanced version of N++. Due to a bug in Windows File Explorer, you cannot open a file located in a long-path folder by double-clicking it in Explorer, nor can you drag and drop such a file into a Notepad++ window to open it. These limitations do not exist in the Total Commander file manager, which has full long-path support.
(I think Microsoft's programmers should probably learn from Total Commander.)

In Windows File Explorer, you can still open files through:
 1. Right-click menu → "Open with…" → choose Notepad++
 2. In Notepad++ main menu → "Open…", then navigate through the long path to open the file.

In N++, many parts of the code need to be modified, mostly involving handling the long-path prefix `L"\\\\?\\"` and adding/removing it correctly.

The main changes I used are:
1. Replace almost all occurrences of `MAX_PATH` with `MAX_UNICODE_PATH`
```
#define MAX_UNICODE_PATH 32767
```

2. Add three common helper functions:
```
bool addLongPathPrefix(std::wstring& path)
{
	if (path.size() > MAX_PATH) {
		static const wchar_t longPrefix[] = L"\\\\?\\";
		const size_t prefixLen = wcslen(longPrefix);

		if (path.compare(0, prefixLen, longPrefix) != 0) {
			path.insert(0, longPrefix);
			return true;
		}
	}
	return false;
}
```
```
bool addLongPathPrefix(wchar_t* path, const size_t bufferSize/* = MAX_UNICODE_PATH*/)
{
	const size_t pathLength = (path != nullptr) ? wcslen(path) : 0;
	if (pathLength > MAX_PATH) {
		static const wchar_t longPrefix[] = L"\\\\?\\";
		const size_t prefixLen = wcslen(longPrefix);

		if (wcsncmp(path, longPrefix, prefixLen) == 0) {
			return true;
		}

		if (pathLength + prefixLen + 1 > bufferSize) {
			return false;
		}

		for (size_t i = pathLength; i != static_cast<size_t>(-1); --i) {
			path[i + prefixLen] = path[i];
		}

		wcsncpy(path, longPrefix, prefixLen);
	}
	return true;
}
```
```
const wchar_t* stripLongPathPrefix(const wchar_t* path)
{
	static const wchar_t longPrefix[] = L"\\\\?\\";
	const size_t prefixLen = wcslen(longPrefix);

	if (path && wcsncmp(path, longPrefix, prefixLen) == 0)
		return path + prefixLen;

	return path;
}
```

3. Then comes a long list of modifications. There are many places that need to be changed, and it's impossible to organize them all at once. Most parts involving file paths need to be checked, such as in the Notepad_plus::doOpen function:
`const rsize_t longFileNameBufferSize = MAX_PATH;` Change to `const rsize_t longFileNameBufferSize = MAX_UNICODE_PATH;`
`wchar_t longFileName[longFileNameBufferSize] = { 0 };` Change to `wchar_t* longFileName = new wchar_t[longFileNameBufferSize] {};` Or `auto longFileName = std::make_unique<wchar_t[]>(longFileNameBufferSize);`

**Comment by xomx** [comment]
@softmgr 
And how about the N++ plugins, did you check that?
I did not but I suppose that if a longpath unaware plugin receives your MAX_UNICODE_PATH name from the N++ via the current plugin API, it will not work or crashes (or could do something worse than that).

**Comment by rdipardo** [comment]
> And how about the N++ plugins, did you check that? I did not but I suppose that if a longpath unaware plugin receives your MAX_UNICODE_PATH name from the N++ via the current plugin API, it will not work or crashes (or could do something worse than that).

There are examples of [C++] and [Pascal] plugins calling the current APIs with over-sized file buffers, so far without issues.

The only use of `MAX_PATH` I can see in the official plugin template is when constructing a hello-world docking dialog:

https://github.com/npp-plugins/plugintemplate/blob/f2a7007dc150d1c448834a2c9df1e9177fcb2c33/src/DockingFeature/DockingDlgInterface.h#L36-L41

This is not code that plugin writers are meant to copy; it's part of a demo that uses an old, DPI-unaware version of `StaticDialog`

Ditching the old `MAX_PATH` limit would at most require updating the docs to warn about the larger buffer limit. The file-path-returning APIs already have a max length parameter in most cases. Unmaintained plugins that set the buffer length to `MAX_PATH` would continue to get 260 max bytes.

Will some plugin users eventually be surprised by truncated file paths? Absolutely. But if that's allowed to hold back a long-overdue, widely requested feature, I worry that N++'s technical debts are already a major obstacle to its continued relevance.

[C++]: https://github.com/rdipardo/nppHTMLTag/blob/4e4f8ffa8c7e35100db8aaa7744c2e0a66a4cbcd/src/LibNppPlugin/PluginBase.cpp#L12
[Pascal]: https://bitbucket.org/rdipardo/delphiplugintemplate/src/893af225d3f52c5c71474303f034a1f8c722bd7d/Source/Units/Common/NppPlugin.pas#lines-140

**Comment by softmgr** [comment]
@xomx 
Currently in N++, all path-related messages (such as `NPPM_GETFULLCURRENTPATH`, `NPPM_GETCURRENTDIRECTORY`, etc.) return a buffer with a maximum size of only 2048 characters:
```
#define CURRENTWORD_MAXLENGTH 2048
```

Even if you change the path constants inside these messages from `MAX_PATH` to `MAX_UNICODE_PATH`, they can still only support path lengths up to 2047 characters. Therefore, every plugin that deals with paths needs to update its calling functions and internal buffer sizes. The `wParam` parameter must be increased to the size of `MAX_UNICODE_PATH` in order to handle long paths properly.

**Comment by mgroeber9110** [comment]
@softmgr 
> Even if you change the path constants inside these messages from `MAX_PATH` to `MAX_UNICODE_PATH`, they can still only support path lengths up to 2047 characters. Therefore, every plugin that deals with paths needs to update its calling functions and internal buffer sizes. The `wParam` parameter must be increased to the size of `MAX_UNICODE_PATH` in order to handle long paths properly.

Would it be an option to allow plugins to be somehow marked as "longname safe" in their metadata, and for those that are not marked use the `GetShortPathName` API function of Windows to try and reduce the size of the filename to `MAX_PATH` by using th 8+3 representations of the path elements?

Only if the path is too long even after this process the call to the plugin would fail.

I have considered making a PR that does this when launching Notepad++ on all paths that would otherwise be too long (because this would be an easy hack for the majority of cases), but if it could be limited to paths passed to plugins, while the main program fully supports long names, this would be even better.

**Comment by donho** [comment]
I tried to open a file with the 399 characters (bytes) path length in the current version of Notepad++:

"C:\tmp\a very very very long path\Show this help message\Show the proxy configuration dialog\Launch GUP with VERSION_VALUE\URL is the url to gain update andor download information\Use URL_PREFIX to verify whether if the download link contain\Older versions are accessible from the\This program is free software you can redistribute\a very very very long path.txt"

The following path is used instead:
`C:\tmp\AVERYV~1\SHOWTH~1\SHOWTH~1\LAUNCH~1\URLIST~1\USEURL~1\OLDERV~1\THISPR~1\AVERYV~1.TXT`

<br/>
<br/>

<img width="874" height="420" alt="Image" src="https://github.com/user-attachments/assets/4adef4e5-2778-4246-9091-09ea249ac6c0" />

<br/>
<br/>

The file was opened via drag & drop and  Open command (with explorer style file dialog).
The 8.3 DOS style path was returned by both [DragQueryFile](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/Notepad_plus.cpp#L4518) and [GetDisplayName](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp#L107) respectively. 

<br/>
<br/>

```
Notepad++ v8.9.2   (64-bit)
Build time: Feb 26 2026 - 14:42:00
Scintilla/Lexilla included: 5.5.8/5.4.6
Boost Regex included: 1_90
pugixml included: 1.15
nlohmann JSON included: 3.12.0
Path: C:\*****\notepad-plus-plus\PowerEditor\visual.net\x64\Debug\Notepad++.exe
Command Line: 
Admin mode: OFF
Local Conf mode: OFF
Cloud Config: OFF
Periodic Backup: ON
Placeholders: OFF
Scintilla Rendering Mode: SC_TECHNOLOGY_DIRECTWRITE (1)
Multi-instance Mode: monoInst
asNotepad: OFF
File Status Auto-Detection: cdEnabledNew (for current file/tab only)
Dark Mode: OFF
Display Info:
    primary monitor: 3840x2160, scaling 225%
    visible monitors count: 1
    installed Display Class adapters: 
        0000: Description - NVIDIA GeForce RTX 3070 Laptop GPU
        0000: DriverVersion - 32.0.15.8132
        0001: Description - Intel(R) UHD Graphics
        0001: DriverVersion - 32.0.101.7077
OS Name: Windows 11 Home (64-bit)
OS Version: 25H2
OS Build: 26200.7840
Current ANSI codepage: 1252
Plugins: none

```

**Comment by softmgr** [comment]
@donho The current version can open converted short paths if they are under 260 characters. But if the path is so long that the short path itself goes over 260 characters, it fails to open.

**Comment by donho** [comment]
@softmgr 
OK, I see

However, as I said:
> The file was opened via drag & drop and  Open command (with explorer style file dialog).
The 8.3 DOS style path was returned by both [DragQueryFile](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/Notepad_plus.cpp#L4518) and [GetDisplayName](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp#L107) respectively. 

How can Notepad++ make DragQueryFile and GetDisplayName return the long path?

**Comment by mgroeber9110** [comment]
@donho 

> How can Notepad++ make DragQueryFile and GetDisplayName return the long path?

As far as I understand, Explorer usually passes shortened names to an executable, even if long paths are enabled system-wide:

https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry#enable-long-paths-in-windows-10-version-1607-and-later

(as long the executable does not declare itself as `longPathAware` in its manifest).

I mostly encounter issues when *other* programs invoke Notepad++ with a non-shorted long path (can be simulated by launching the program from the command line with a full path) - in my case, this is WinSCP with Notepad++ as the preferred editor when I double-click on a log file (because WinSCP includes the full server-side path for the temporary copy that is downloaded).

**Comment by softmgr** [comment]
> @softmgr 
> OK, I see
> 
> However, as I said:
> > The file was opened via drag & drop and  Open command (with explorer style file dialog).
> The 8.3 DOS style path was returned by both [DragQueryFile](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/Notepad_plus.cpp#L4518) and [GetDisplayName](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp#L107) respectively. 
> 
> How can Notepad++ make DragQueryFile and GetDisplayName return the long path?

The DragQueryFile function does not support long paths, and the program may not even receive the WM_DROPFILES message at all. It is necessary to switch to the IDropTarget interface and use RegisterDragDrop to register for drag-and-drop support.

**Comment by softmgr** [comment]
@donho 
I attempted to write some OLE drag-and-drop code. The details are as follows:
 1. Dragging and dropping files or folders onto the window is correctly recognized and opens them.
 2. Dragging and dropping Unicode or Ansi text onto the window inserts the text into the editor (this functionality was not present in previous versions).
 3. Dragging and dropping files, folders, or text onto the Scintilla editor area still uses the old WM_DROPFILES message handling because I tried removing the WM_DROPFILES message, but the Scintilla editor area would no longer accept file and folder drag-and-drop. This issue may need to be addressed by an update from the Scintilla team.
 4. If Windows UAC is enabled and the application is running with administrator privileges, the OLE drag-and-drop functionality fails and reverts to the old WM_DROPFILES message handling (which cannot support long paths because WM_DROPFILES only supports short paths).

Here is the code I added, please refer to it:


//Notepad_plus.h

```
class NppDropTarget : public IDropTarget
{
public:
	NppDropTarget(Notepad_plus* pNpp, HWND hwnd) :
		_refCount(1L), _pNpp(pNpp), _hNppWnd(hwnd)
	{
	}

	// IUnknown
	HRESULT __stdcall QueryInterface(REFIID riid, void** ppv) override
	{
		if (riid == IID_IUnknown || riid == IID_IDropTarget)
		{
			*ppv = this;
			AddRef();
			return S_OK;
		}
		*ppv = nullptr;
		return E_NOINTERFACE;
	}

	ULONG __stdcall AddRef() override
	{
		return InterlockedIncrement(&_refCount);
	}

	ULONG __stdcall Release() override
	{
		LONG count = InterlockedDecrement(&_refCount);
		if (count == 0)
			delete this;
		return count;
	}

	// IDropTarget
	HRESULT __stdcall DragEnter(
		IDataObject* pDataObj,
		DWORD,
		POINTL,
		DWORD* pdwEffect) override
	{
		if (!pDataObj)
		{
			*pdwEffect = DROPEFFECT_NONE;
			return E_POINTER;
		}
		*pdwEffect = DROPEFFECT_COPY;
		return S_OK;
	}

	HRESULT __stdcall DragOver(
		DWORD,
		POINTL,
		DWORD* pdwEffect) override
	{
		*pdwEffect = DROPEFFECT_COPY;
		return S_OK;
	}

	HRESULT __stdcall DragLeave() override
	{
		return S_OK;
	}

	HRESULT __stdcall Drop(
		IDataObject* pDataObj,
		DWORD,
		POINTL pt,
		DWORD* pdwEffect) override
	{
		handleDrop(pDataObj, pt);
		*pdwEffect = DROPEFFECT_COPY;
		return S_OK;
	}

	LONG GetRefCount() const { return _refCount; }

private:
	void handleDrop(IDataObject* pDataObj, POINTL pt);

private:
	LONG _refCount = 0;
	Notepad_plus* _pNpp = nullptr;
	HWND _hNppWnd = nullptr;
};

class Notepad_plus final
{
friend class Notepad_plus_Window;
friend class FileManager;
friend class NppDropTarget;  // added

public:
	Notepad_plus();
	~Notepad_plus();

	bool regOleDragDrop(HWND hwnd);  // added
	void unregOleDragDrop(HWND hwnd);  // added

private:
	NppDropTarget* _dropTarget = nullptr;  // added
	HRESULT _hrOle = E_FAIL;  // added

};
```



//Notepad_plus.cpp

```
#include <shlobj.h>  // added



void NppDropTarget::handleDrop(IDataObject* pDataObj, POINTL pt)  // added
{
	//if (!_pNpp)
	//	return;

	//
	// ===== Switch to view based on mouse position =====
	//
	POINT p{ pt.x, pt.y };
	::ScreenToClient(_hNppWnd, &p);
	HWND hWin = ::ChildWindowFromPointEx(_hNppWnd, p, CWP_SKIPINVISIBLE);
	if (hWin)
	{
		if ((_pNpp->_mainEditView.getHSelf() == hWin) || (_pNpp->_mainDocTab.getHSelf() == hWin))
			_pNpp->switchEditViewTo(MAIN_VIEW);
		else if ((_pNpp->_subEditView.getHSelf() == hWin) || (_pNpp->_subDocTab.getHSelf() == hWin))
			_pNpp->switchEditViewTo(SUB_VIEW);
		//else
			// do not change the current Notepad++ edit-view
	}

	//
	// ===== Try IShellItemArray first =====
	//
	IShellItemArray* pItemArray = nullptr;

	HRESULT hr = ::SHCreateShellItemArrayFromDataObject(
		pDataObj,
		IID_PPV_ARGS(&pItemArray));

	if (SUCCEEDED(hr) && pItemArray)
	{
		DWORD count = 0;
		pItemArray->GetCount(&count);

		std::vector<std::wstring> folderPaths;
		std::vector<std::wstring> filePaths;
		std::vector<std::wstring_view> allPaths;
		folderPaths.reserve(count);
		filePaths.reserve(count);
		allPaths.reserve(count);

		for (DWORD i = 0; i < count; ++i)
		{
			IShellItem* pItem = nullptr;
			if (SUCCEEDED(pItemArray->GetItemAt(i, &pItem)))
			{
				PWSTR pszPath = nullptr;

				if (SUCCEEDED(pItem->GetDisplayName(
					SIGDN_FILESYSPATH,
					&pszPath)))
				{
					std::wstring path(pszPath);
					addLongPathPrefix(path);
					if (doesPathExist(path.c_str()))
					{
						if (doesDirectoryExist(path.c_str()))
						{
							if (path.back() != L'\\')
								path += L'\\';
							folderPaths.emplace_back(std::move(path));
							allPaths.emplace_back(folderPaths.back());
						}
						else
						{
							filePaths.emplace_back(std::move(path));
							allPaths.emplace_back(filePaths.back());
						}
					}
					::CoTaskMemFree(pszPath);
				}

				pItem->Release();
			}
		}

		pItemArray->Release();

		NppParameters& nppParam = NppParameters::getInstance();
		bool isOldMode = nppParam.getNppGUI()._isFolderDroppedOpenFiles;

		if (isOldMode || folderPaths.size() == 0) // old mode or new mode + only files
		{
			BufferID lastOpened = BUFFER_INVALID;
			for (std::wstring_view path : allPaths)
			{
				BufferID test = _pNpp->doOpen(std::wstring(path));
				if (test != BUFFER_INVALID)
					lastOpened = test;
			}

			if (lastOpened != BUFFER_INVALID)
			{
				_pNpp->switchToFile(lastOpened);
			}
		}
		else if (!isOldMode && (folderPaths.size() != 0 && filePaths.size() != 0)) // new mode && both folders & files
		{
			// display error & do nothing
			_pNpp->_nativeLangSpeaker.messageBox("DroppingFolderAsProjectModeWarning",
				_hNppWnd,
				L"You can only drop files or folders but not both, because you're in dropping Folder as Project mode.\rYou have to enable \"Open all files of folder instead of launching Folder as Workspace on folder dropping\" in \"Default Directory\" section of Preferences dialog to make this operation work.",
				L"Invalid action",
				MB_OK | MB_APPLMODAL);
		}
		else if (!isOldMode && (folderPaths.size() != 0 && filePaths.size() == 0)) // new mode && only folders
		{
			// process new mode
			std::wstring emptyStr;
			_pNpp->launchFileBrowser(folderPaths, emptyStr);
		}
	}
	else
	{
		//
		// ===== Processing Unicode Text =====
		//
		std::wstring textDropped;  // Drag and drop text
		{
			FORMATETC fmtUnicode = { CF_UNICODETEXT, nullptr, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };
			STGMEDIUM stg{};
			if (SUCCEEDED(pDataObj->GetData(&fmtUnicode, &stg)))
			{
				wchar_t* pText = static_cast<wchar_t*>(::GlobalLock(stg.hGlobal));
				if (pText)
				{
					textDropped = pText;
					::GlobalUnlock(stg.hGlobal);
				}
				::ReleaseStgMedium(&stg);
			}
		}

		//
		// ===== fallback ANSI Text =====
		//
		if (textDropped.empty())
		{
			FORMATETC fmtAnsi = { CF_TEXT, nullptr, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };
			STGMEDIUM stgAnsi{};
			if (SUCCEEDED(pDataObj->GetData(&fmtAnsi, &stgAnsi)))
			{
				char* pText = static_cast<char*>(::GlobalLock(stgAnsi.hGlobal));
				if (pText)
				{
					textDropped = string2wstring(pText);
					::GlobalUnlock(stgAnsi.hGlobal);
				}
				::ReleaseStgMedium(&stgAnsi);
			}
		}

		// Insert drag and drop text
		if (!textDropped.empty())
		{
			UINT cp = static_cast<UINT>(_pNpp->_pEditView->execute(SCI_GETCODEPAGE));
			std::string replaceTextA = wstring2string(textDropped, cp);
			if (!replaceTextA.empty())
				_pNpp->_pEditView->execute(SCI_REPLACESEL, 0, reinterpret_cast<LPARAM>(replaceTextA.c_str()));
		}
	}

	// Put Notepad_plus to forefront
	// May not work for Win2k, but OK for lower versions
	// Note: how to drop a file to an iconic window?
	// Actually, it is the Send To command that generates a drop.
	if (::IsIconic(_hNppWnd))
	{
		::ShowWindow(_hNppWnd, SW_RESTORE);
	}
	::SetForegroundWindow(_hNppWnd);
}
```


//NppBigSwitch.cpp

```
LRESULT Notepad_plus_Window::runProc(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam)
{
	switch (message)
	{
		case WM_CREATE:
		{
			try
			{
				NppDarkMode::setDarkTitleBar(hwnd);
				NppDarkMode::autoSubclassWindowMenuBar(hwnd);
				NppDarkMode::autoSubclassCtlColor(hwnd);

				_notepad_plus_plus_core._pPublicInterface = this;
				LRESULT lRet = _notepad_plus_plus_core.init(hwnd);
				_notepad_plus_plus_core.regOleDragDrop(hwnd);  // added

				if (NppDarkMode::isEnabled() && NppDarkMode::isExperimentalSupported())
				{
					// Inform application of the frame change.
					::SetWindowPos(hwnd, nullptr, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOSIZE | SWP_NOZORDER | SWP_FRAMECHANGED);
				}

				SetOSAppRestart();

				return lRet;
			}
			catch (std::exception& ex)
			{
				::MessageBoxA(hwnd, ex.what(), "Exception On WM_CREATE", MB_OK);
				exit(-1);
			}
			break;
		}
		default:
		{
			return _notepad_plus_plus_core.process(hwnd, message, wParam, lParam);
		}
	}
}

bool Notepad_plus::regOleDragDrop(HWND hwnd)  // added
{
	if (!_dropTarget && !_isAdministrator) {
		_hrOle = ::OleInitialize(nullptr);

		if (SUCCEEDED(_hrOle)) {
			_dropTarget = new NppDropTarget(this, hwnd);
			HRESULT s = ::RegisterDragDrop(hwnd, _dropTarget);
			if (SUCCEEDED(s))
				return true;
		}
	}
	return false;
}

void Notepad_plus::unregOleDragDrop(HWND hwnd)  // added
{
	if (_dropTarget && SUCCEEDED(_hrOle)) {
		::RevokeDragDrop(hwnd);

		if (_dropTarget->GetRefCount() == 1L)
			_dropTarget->Release();
		_dropTarget = nullptr;

		::OleUninitialize();
	}
}

LRESULT Notepad_plus::process(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam)
{
	LRESULT result = FALSE;
	NppParameters& nppParam = NppParameters::getInstance();

	switch (message)
	{
		case WM_DESTROY:
		{
			// ...

			unregOleDragDrop(hwnd);  // added
			killAllChildren();
			::PostQuitMessage(0);
			_pPublicInterface->gNppHWND = NULL;
			return TRUE;
		}
	}
	
	return result;
}
```

**Comment by nyamatongwe** [comment]
[Moving discussion from #17562]

> As for the WM_DROPFILES message ... Therefore, the current choice is to use OLE drag-and-drop

Other applications are using `WM_DROPFILES` so Scintilla must remain compatible with them.

If this change is not compatible with other applications then it may be reasonable to add an enabling API, possibly as a flag to `SCI_SETDRAGDROPENABLED`.

> Therefore, the LPDATAOBJECT pIDataSource variable recognized by Scintilla must be passed directly to the main window for processing (instead of sending the file name).

*must*? I think not. The issue here is to add support for long paths to Notepad++. Adding support for other features is out of scope and will require more design and discussion. A proposal should be posted to the [Scintilla mailing list](https://groups.google.com/g/scintilla-interest).

The simplest way forward appears to me to implement `SCN_URIDROPPED` in the Win32 platform layer.
